core/src/main/java/io/grpc/internal/SpiffeUtil.java

```diff
@@ -0,0 +1,186 @@
+
+    Map<String, List<X509Certificate>> trustBundleMap = new HashMap<>();
+    Map<String, Long> sequenceNumbers = new HashMap<>();
+    for (String trustDomainName : trustDomainsNode.keySet()) {
+      Map<String, ?> domainNode = JsonUtil.getObject(trustDomainsNode, trustDomainName);
+      if (domainNode == null || domainNode.size() == 0) {
+        trustBundleMap.put(trustDomainName, Collections.emptyList());
+        continue;
+      }
+      Long sequenceNumber = JsonUtil.getNumberAsLong(domainNode, "sequence_number");
+      sequenceNumbers.put(trustDomainName, sequenceNumber);
+      List<Map<String, ?>> keysNode = JsonUtil.getListOfObjects(domainNode, "keys");
+      if (keysNode == null || keysNode.size() == 0) {
+        trustBundleMap.put(trustDomainName, Collections.emptyList());
+        continue;
+      }
+      List<X509Certificate> roots = new ArrayList<>();
+      for (Map<String, ?> keyNode : keysNode) {
+        String kid = JsonUtil.getString(keyNode, "kid");
+        if (kid != null && !kid.equals("")) {
+          log.log(Level.SEVERE, String.format("'kid' parameter must not be set but value '%s' "
+              + "found. Skipping certificate loading for trust domain '%s'.", kid,
+              trustDomainName));
+          break;
+        }
+        String use = JsonUtil.getString(keyNode, "use");
+        if (use == null || !use.equals(USE_PARAMETER_VALUE)) {
+          log.log(Level.SEVERE, String.format("'use' parameter must be '%s' but '%s' found. "
+              + "Skipping certificate loading for trust domain '%s'.", USE_PARAMETER_VALUE, use,
+              trustDomainName));
+          break;
+        }
+        String rawCert = JsonUtil.getString(keyNode, "x5c");
+        if (rawCert == null) {
+          break;
+        }
+        InputStream stream = new ByteArrayInputStream(rawCert.getBytes(StandardCharsets.UTF_8));
+        try {
+          Collection<? extends Certificate> certs = CertificateFactory.getInstance("X509").
+              generateCertificates(stream);
+          if (certs.size() > 0){
+            roots.add(certs.toArray(new X509Certificate[0])[0]);
+          }
+        } catch (CertificateException e) {
+          log.log(Level.SEVERE, String.format("Certificate for domain %s can't be parsed.",
+              trustDomainName), e);
+        }
+      }
+      trustBundleMap.put(trustDomainName, roots);
+    }
+    return new TrustBundle(sequenceNumbers, trustBundleMap);
+  }
+
+  // Will be merged with other PR
+  public static class SpiffeId {
+
+    private final String trustDomain;
+    private final String path;
+
+    private SpiffeId(String trustDomain, String path) {
+      this.trustDomain = trustDomain;
+      this.path = path;
+    }
+
+    public String getTrustDomain() {
+      return trustDomain;
+    }
+
+    public String getPath() {
+      return path;
+    }
+  }
+
+  public static class TrustBundle {
+
+    private final Map<String, Long> sequenceNumbers;
```

Comment:
Thanks for explaining. Are we going to use the sequence numbers for anything? If not, should we drop them?

Reply:
IIUC the ideas behind the spec, we can use it for logging (because it indicates a kind of version of that particular TD  entry in the file) but I don't think we'll be making any decisions based on these numbers (like 'Hey let's reject this entry because current sequence number is less than the previous one).

Reply:
Thanks for the context. If they're optional and we don't need to use them for anything, can we omit them to simplify the code?